File: packages/ios/Sources/Rejourney/Recording/RejourneyNetworkEventFilter.swift
/**
 * Copyright 2026 Rejourney
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Foundation

enum RejourneyNetworkEventFilter {
    private static let lock = NSLock()
    private static let internalPathPrefixes = [
        "/api/sdk/config",
        "/api/ingest",
        "/upload/artifacts"
    ]
    private static var apiBaseURLString = normalizeBaseURLString("https://api.rejourney.co")

    static func configure(apiURLString: String) {
        lock.lock()
        apiBaseURLString = normalizeBaseURLString(apiURLString)
        lock.unlock()
    }

    static func shouldIgnore(url: URL) -> Bool {
        let absoluteString = url.absoluteString
        let configuredBase = currentAPIBaseURLString()
        if !configuredBase.isEmpty && absoluteString.contains(configuredBase) {
            return true
        }

        return shouldIgnore(path: url.path.isEmpty ? "/" : url.path)
    }

    static func shouldIgnore(details: [String: Any]) -> Bool {
        if let urlString = details["url"] as? String,
           let url = URL(string: urlString),
           shouldIgnore(url: url) {
            return true
        }

        if let path = details["urlPath"] as? String {
            return shouldIgnore(path: path)
        }

        return false
    }

    private static func shouldIgnore(path: String) -> Bool {
        internalPathPrefixes.contains { prefix in
            path == prefix || path.hasPrefix("\(prefix)/")
        }
    }

    private static func currentAPIBaseURLString() -> String {
        lock.lock()
        defer { lock.unlock() }
        return apiBaseURLString
    }

    private static func normalizeBaseURLString(_ value: String) -> String {
        value.trimmingCharacters(in: .whitespacesAndNewlines)
            .trimmingCharacters(in: CharacterSet(charactersIn: "/"))
    }
}

File: packages/react-native/android/src/main/java/com/rejourney/recording/RejourneyNetworkEventFilter.kt
/**
 * Copyright 2026 Rejourney
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.rejourney.recording

import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import java.net.URI

object RejourneyNetworkEventFilter {
    private val internalPathPrefixes = listOf(
        "/api/sdk/config",
        "/api/ingest",
        "/upload/artifacts"
    )

    @Volatile
    private var apiBaseUrl: String = normalizeBaseUrl("https://api.rejourney.co")

    fun configure(apiUrl: String?) {
        apiBaseUrl = normalizeBaseUrl(apiUrl ?: "https://api.rejourney.co")
    }

    fun shouldIgnore(url: HttpUrl): Boolean {
        val configuredBase = apiBaseUrl
        if (configuredBase.isNotEmpty() && url.toString().contains(configuredBase)) {
            return true
        }
        return shouldIgnorePath(url.encodedPath)
    }

    fun shouldIgnore(url: String?): Boolean {
        if (url.isNullOrBlank()) return false
        val configuredBase = apiBaseUrl
        if (configuredBase.isNotEmpty() && url.contains(configuredBase)) {
            return true
        }
        return shouldIgnorePath(pathForUrl(url))
    }

    fun shouldIgnore(details: Map<String, Any>): Boolean {
        val url = details["url"] as? String
        if (shouldIgnore(url)) return true

        val path = details["urlPath"] as? String
        return path != null && shouldIgnorePath(path)
    }

    private fun shouldIgnorePath(path: String): Boolean {
        return internalPathPrefixes.any { prefix ->
            path == prefix || path.startsWith("$prefix/")
        }
    }

    private fun pathForUrl(value: String): String {
        value.toHttpUrlOrNull()?.let { return it.encodedPath }
        return try {
            URI(value).rawPath ?: value.substringBefore('?')
        } catch (_: Exception) {
            value.substringBefore('?')
        }
    }

    private fun normalizeBaseUrl(value: String): String {
        return value.trim().trimEnd('/')
    }
}
